### PR TITLE
feed: top libraries

### PR DESCRIPTION
**Pull request by raymondng**
Baby steps. For today just one simple endpoint based on `mostMembersSince`. 

Need a lot more work but it's functional and more importantly RSS-compliant (AFAIK, new to the format myself)

Note I have added a new top level `/feeds` for the RSS endpoints. 

Plan is to publish this and then see if it works properly in feed reader, etc. And go from there. Feedback appreciated.

@andrewconner @eishay

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 1]
```diff
@@ -55,6 +63,51 @@ class LibraryController @Inject() (
...
+
+    val libIds = tops.map(_._1)
+    val libMap = db.readOnlyReplica { implicit request =>
+      libraryRepo.getLibraries(libIds.toSet)
+    }
+    val libraries = libIds.map(libMap(_))
+    val ownerIds = CollectionHelpers.dedupBy(libraries.map(_.ownerId))(id => id)
+    val ownerMap = db.readOnlyMaster { implicit request => userRepo.getUsers(ownerIds) }
+    val formatter = DateTimeFormat.forPattern("E, d MMM y HH:mm:ss Z") // rfc822
+    val prefix = if (mode == Mode.Dev) "http://dev.ezkeep.com:9000" else "https://www.kifi.com"
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: maybe use `FortyTwoConfig#applicationBaseUrl` ?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 2]
```diff
@@ -55,6 +63,51 @@ class LibraryController @Inject() (
...
+    val items = tops map {
+      case (libId, count) =>
+        val lib = libMap(libId)
+        val owner = ownerMap(lib.ownerId)
+        <item>
+          <title>{ lib.slug.value }</title>
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: why not `lib.name`?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 1 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Cool. Didn't know -- I was going to add something like this.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 3]
```diff
@@ -55,6 +63,51 @@ class LibraryController @Inject() (
...
+          <description>{ lib.description.getOrElse("") }</description>
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: See `LibraryCommander#libraryMetaTags`
How about using instead

``` scala
library.description.getOrElse(s"${owner.fullName}'s ${library.name} Kifi Library")
```

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 2 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: that's a good idea. thx.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 4]
```diff
@@ -55,6 +63,51 @@ class LibraryController @Inject() (
...
+        </channel>
+      </rss>
+
+    val elems = Enumerator.enumerate(rss)
+    val toBytes = Enumeratee.map[Node] { n => n.toString.getBytes }
+    val header = Enumerator("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n".getBytes)
+    val combined = header.andThen(elems &> toBytes)
+    log.info(s"[getTopLibraries] published #${libIds.size} libraries. ${libIds.take(20)} ...")
+    Result(
+      header = ResponseHeader(200, Map(CONTENT_TYPE -> "application/xml")),
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: http://tools.ietf.org/id/draft-nottingham-rss-media-type-00.txt

``` application/rss+xml```
```

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 3 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Sounds good.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 4 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Good catch. Feed readers processed it without issues, thought I am compliant, but guess not.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 5]
```diff
@@ -55,6 +63,51 @@ class LibraryController @Inject() (
...
+        val owner = ownerMap(lib.ownerId)
+        <item>
+          <title>{ lib.slug.value }</title>
+          <description>{ lib.description.getOrElse("") }</description>
+          <link>{ s"$prefix${Library.formatLibraryPath(owner.username, owner.externalId, lib.slug)}" }</link>
+          <pubDate>{ lib.createdAt.toString(formatter) }</pubDate>
+        </item>
+    }
+    val rss =
+      <rss>
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: not 

``` xml
<rss version="2.0">
```

http://en.wikipedia.org/wiki/RSS#Example
?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 6]
```diff
@@ -55,6 +63,51 @@ class LibraryController @Inject() (
...
+    val prefix = if (mode == Mode.Dev) "http://dev.ezkeep.com:9000" else "https://www.kifi.com"
+    val items = tops map {
+      case (libId, count) =>
+        val lib = libMap(libId)
+        val owner = ownerMap(lib.ownerId)
+        <item>
+          <title>{ lib.slug.value }</title>
+          <description>{ lib.description.getOrElse("") }</description>
+          <link>{ s"$prefix${Library.formatLibraryPath(owner.username, owner.externalId, lib.slug)}" }</link>
+          <pubDate>{ lib.createdAt.toString(formatter) }</pubDate>
```
Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: GUID helps scrapers a lot. We can use the library external id. E.e. 

``` scala
<guid>{ lib.externalid.toString }</guid>
```

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 5 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: I was following Google doc (which does not include the version) but I can add it.

**Comment by eishay** [comment]
lets use the `author` tag http://cyber.law.harvard.edu/rss/rss.html#hrelementsOfLtitemgt with the `owner.fullName`

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 6 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Interesting. I actually looked into GUID a bit and thought it's not necessary -- I checked a few top sites ... they seem to just dup it with link or drop it entirely. I can add it though.

Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: I used the GUID at linkedin when i built the scraper to dedup articles.
Think about renaming the library, updating the slug or moving the owner.
The GUID will tell the scraper its still the same web resource.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 7]
```diff
@@ -55,6 +64,49 @@ class LibraryController @Inject() (
   implicit val config: PublicIdConfiguration)
     extends UserActions with LibraryAccessActions with ShoeboxServiceController {
 
+  def getTopLibraries() = MaybeUserAction { request =>
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:66]: Why `MaybeUserAction`?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 8]
```diff
@@ -55,6 +64,49 @@ class LibraryController @Inject() (
...
+    val tops = db.readOnlyReplica { implicit request =>
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Shadowing `request`

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala [hunk 9]
```diff
@@ -55,6 +64,49 @@ class LibraryController @Inject() (
...
+    }
+
+    val libIds = tops.map(_._1)
+    val libMap = db.readOnlyReplica { implicit request =>
+      libraryRepo.getLibraries(libIds.toSet)
+    }
+    val libraries = libIds.map(libMap(_))
+    val ownerIds = CollectionHelpers.dedupBy(libraries.map(_.ownerId))(id => id)
+    val ownerMap = db.readOnlyMaster { implicit request => userRepo.getUsers(ownerIds) }
+    val formatter = DateTimeFormat.forPattern("E, d MMM y HH:mm:ss Z") val items = tops map {
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Two lines combined?

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 8 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Ah will make it clearer

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 9 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Weird. I am sure I didn't do this ... will fix.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 7 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:66]: Knowing that a user is visiting might be helpful, but it's not necessary. Are you concerned about overhead?

Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:66]: It's not being used.

Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala:66]: ah that will be fixed shortly :)

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala (hunk 6 above)
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: Ah we have an in-house expert! You should have told me earlier so I can bug you instead :)
No problem -- will add GUID. BTW we have no external Id, only public Id -- is that what you had in mind?

Comment (eishay) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/website/LibraryController.scala]: That would work as well.

**Comment by andrewconner** [comment]
Wahoo. Let's get this added to the marketing homepage.

**Comment by eishay** [comment]
Sure, get it out. 
I would like to get some stats on it quickly e.g.:
- How frequently its pulled (maybe we should cache) 
- Adding tracking to the urls @stkem can help with how to do it

**Comment by raymondng** [comment]
Alright. On it!

**Pull request merged**